addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10Interface.java

```diff
@@ -0,0 +1,800 @@
+                    // As indicated above, some controllers send the address in one transmission and the function in a
+                    // second transaction.
+                    // Check if we have gotten an address in the transmission and if not use lastAddresses if they
+                    // are not null
+                    if (addresses.isEmpty()) {
+                        // No addresses were sent in this transmission, see if we can use lastAddresses
+                        if (lastAddresses == null) {
+                            log.warn("cm11a received a command but the transmission didn't include an address.");
+                            continue;
+                        } else {
+                            addresses = lastAddresses;
+                            log.warn(
+                                    "cm11a received a command without any addresses. Addresses from a prior reception are available and will be used.");
+                        }
+                    }
+                }
+
+                // Every time we get a function it is the end of the transmission and we can bundle the data into an
+                // X10ReceivedData object.
+                X10ReceivedData rd = new X10ReceivedData(addresses.toArray(new String[0]), command, dims);
+                rcvData.add(rd);
+                log.debug("cm11a: Added received data to queue: " + rd.toString());
+                // reset the data objects for the next thing in the buffer, if any
+                command = X10ReceivedData.X10COMMAND.UNDEF;
+                // Collections.copy(lastAddresses, addresses);
+                lastAddresses = addresses;
+                addresses = new ArrayList<>();
+                dims = 0;
+            }
+
+            mask = mask >> 1;
+        }
+
+        // Done processing buffer from cm11a. Notify interested parties about the data
+        for (X10ReceivedData rd : rcvData) {
+            log.debug("cm11a: Converted received data to human form: " + rd.toString());
+            notifyReceiveListeners(rd);
+        }
+    }
+
+    /**
+     * Called by classes that want to be notified when data has been received from the cm11a
+     *
+     * @param listener The class to be called
+     */
+    public void addReceivedDataListener(ReceivedDataListener listener) {
+        receiveListeners.add(listener);
+    }
+
+    /**
+     * Called to notify classes that data has been received
+     *
+     * @param rcv
+     */
+    private void notifyReceiveListeners(X10ReceivedData rcv) {
+
+        for (ReceivedDataListener rl : receiveListeners) {
+            rl.receivedX10Data(rcv);
+        }
+    }
+
+    /**
+     * Disconnect from hardware
+     */
+    public void disconnect() {
+        killThread = true;
+        try {
+            this.wait((IO_RECONNECT_INTERVAL + IO_PORT_OPEN_TIMEOUT) * 2);
+        } catch (InterruptedException e) {
+            // Nothing to do.
```

Comment:
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/